pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -184,3 +164,39 @@ func (n Node) hasCompatibleResources(resourceList v1.ResourceList, it cloudprovi
 	}
 	return true
 }
+
+func (n Node) hasCompatibleNodeAffinity(affinity *v1.Affinity, it cloudprovider.InstanceType) bool {
+	if affinity == nil || affinity.NodeAffinity == nil || affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return true
+	}
+
+	for _, term := range affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		if term.MatchExpressions != nil {
+			if !compatibleInstanceType(v1alpha5.NewRequirements(term.MatchExpressions...), it) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func compatibleInstanceType(requirements v1alpha5.Requirements, it cloudprovider.InstanceType) bool {
```

Comment:
Your node object already has `Constraints.Requirements`, and you can avoid the static.
```suggestion
func (n *Node) hasCompatibleInstanceType(it cloudprovider.InstanceType) bool {
```

Reply:
It's also used with a generated set of requirements from `hasCompatibleNodeAffinity`